modules/core/src/test/java/org/apache/ignite/internal/processors/security/cluster/ClusterStatePermissionTest.java

```diff
@@ -0,0 +1,344 @@
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String instanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(instanceName);
+
+        if (!cfg.isClientMode()) {
+            cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(new DataRegionConfiguration().setPersistenceEnabled(persistence)));
+        }
+
+        SecurityPermission[] srvPerms = EMPTY_PERMS;
+        SecurityPermission[] clientPerms = EMPTY_PERMS;
+
+        if (initiator == Initiator.SERVER)
+            srvPerms = F.concat(srvPerms, permissions);
+        else if (initiator == Initiator.CLIENT)
+            srvPerms = permissions;
+        else if (initiator == Initiator.THIN_CLIENT)
+            clientPerms = permissions;
+        else if (initiator == Initiator.REMOTE_CONTROL) {
+            clientPerms = permissions;
+
+            cfg.setConnectorConfiguration(new ConnectorConfiguration());
+        }
+
+        SecurityPermissionSetBuilder secBuilder = create().defaultAllowAll(false)
+            .appendSystemPermissions(F.concat(srvPerms, JOIN_AS_SERVER));
+
+        TestSecurityPluginProvider secPlugin = new TestSecurityPluginProvider(
+            instanceName,
+            "",
+            secBuilder.build(),
+            false,
+            new TestSecurityData(
+                "client",
+                "",
+                create().defaultAllowAll(false).appendSystemPermissions(clientPerms).build(),
+                new Permissions()
+            )
+        );
+
+        cfg.setPluginProviders(secPlugin);
+
+        cfg.setClusterStateOnStart(INACTIVE);
+
+        return cfg;
+    }
+
+    /**
+     * Tests that all state changes are allowed with the permission.
+     */
+    @Test
+    public void testActivationDeactivationAllowed() throws Exception {
+        permissions = F.asArray(ADMIN_CLUSTER_STATE);
+
+        doTestChangeState(
+            F.asArray(ACTIVE_READ_ONLY, INACTIVE, ACTIVE_READ_ONLY, ACTIVE, INACTIVE, ACTIVE, ACTIVE_READ_ONLY),
+            true
+        );
+    }
+
+    /**
+     * Tests that the activation is not allowed without the permission.
+     */
+    @Test
+    public void testActivationNotAllowed() throws Exception {
+        doTestChangeState(F.asArray(ACTIVE), false);
+    }
+
+    /**
+     * Tests that the deactivation is not allowed without the permission.
+     */
+    @Test
+    public void testDeactivationNotAllowed() throws Exception {
+        startAllAllowedNode();
+
+        doTestChangeState(F.asArray(INACTIVE), false);
+    }
+
+    /**
+     * Tests that same-state change is not allowed without the permission.
+     */
+    @Test
+    public void testSameStateNotAllowedWithoutPermission() throws Exception {
+        startAllAllowedNode();
```

Comment:
I removed startAllAllowedNode() call here and above and all tests passed. Do we need it?

Reply:
If you remove `startAllAllowedNode()` from `testSameStateNotAllowedWithoutPermission()`, it becomes `testActivationNotAllowed()`, But that it other test. `testSameStateNotAllowedWithoutPermission()` sets ACTIVE stame from node, where it is allowed. And check that a SecurityException is throw from other node, where the permission is not set even if you pass same state. `testActivationNotAllowed()` checks that activation in not possible from INACTIVE state without the permission.